crates/hulk_widgets/src/completion_edit.rs

```diff
@@ -0,0 +1,308 @@
+        let output = self.show(ui, &mut state, show_value);
+        state.save(ui.ctx(), self.id);
+        output
+    }
+
+    fn show(
+        &mut self,
+        ui: &mut Ui,
+        state: &mut CompletionEditState,
+        show_value: impl Fn(&mut Ui, bool, &T) -> Response,
+    ) -> Response {
+        let matching_items = ui.memory_mut(|writer| {
+            let cache = writer.caches.cache::<CachedMatcherSearch>();
+            cache.get((self.selected, self.suggestions))
+        });
+        let popup_id = self.id.with("popup");
+        let is_popup_open = ui.memory(|reader| reader.is_popup_open(popup_id));
+
+        let (pressed_up, pressed_down) = if is_popup_open {
+            ui.input_mut(|reader| {
+                (
+                    reader.consume_key(Modifiers::NONE, Key::ArrowUp)
+                        || reader.consume_key(Modifiers::SHIFT, Key::Tab),
+                    reader.consume_key(Modifiers::NONE, Key::ArrowDown)
+                        || reader.consume_key(Modifiers::NONE, Key::Tab),
+                )
+            })
+        } else {
+            (false, false)
+        };
+        let TextEditOutput {
+            mut response,
+            state: text_edit_state,
+            ..
+        } = ui
+            .scope(|ui| match state.user_state {
+                UserState::Typing => TextEdit::singleline(self.selected)
+                    .hint_text("Search")
+                    .show(ui),
+                UserState::Selecting { index } => {
+                    let mut active_stroke = ui.visuals().widgets.active.bg_stroke;
+                    active_stroke.color = ui.visuals().warn_fg_color;
+                    ui.visuals_mut().widgets.inactive.bg_stroke = active_stroke;
+
+                    let mut selected = matching_items
+                        .get(index)
+                        .map(|(_, value)| value.clone())
+                        .unwrap_or_default();
+                    let output = TextEdit::singleline(&mut selected)
+                        .text_color(Color32::GRAY)
+                        .hint_text("Search")
+                        .show(ui);
+                    set_cursor(
+                        ui.ctx(),
+                        &output.response,
+                        output.state.clone(),
+                        selected.len(),
+                        selected.len(),
+                    );
+                    if output.response.changed() {
+                        *self.selected = selected;
+                        state.user_state = UserState::Typing;
+                    }
+                    output
+                }
+            })
+            .inner;
+
+        if response.changed() {
+            state.typed_since_focused = true;
+        }
+        if response.gained_focus() {
+            // Select all
+            set_cursor(ui.ctx(), &response, text_edit_state, 0, self.selected.len());
```

Comment:
Last time I checked, this did not work. Pressing CTRL+P in twix focuses the panel selector, but did not select the text inside.

Reply:
yes, `gained_focus` seems to be broken in egui, now implemented it myself in the widget